janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/StaticArrayEntryList.java

```diff
@@ -398,11 +401,89 @@ private static <E,D> EntryList of(Iterable<E> elements, StaticArrayEntry.GetColV
+    }
+
+    public static <E> void supplyEntryList(ChunkedJobDefinition<Iterator<E>, EntryListComputationContext, EntryList> chunkedJobDefinition,
+                                           StaticArrayEntry.GetColVal<E,StaticBuffer> getter,
+                                           ExecutorService executorService) {
+        supplyEntryList(chunkedJobDefinition, getter, StaticArrayEntry.StaticBufferHandler.INSTANCE, executorService);
+    }
+
+    private static <E,D> void supplyEntryList(ChunkedJobDefinition<Iterator<E>, EntryListComputationContext, EntryList> chunkedJobDefinition,
+                                              StaticArrayEntry.GetColVal<E,D> getter,
+                                              StaticArrayEntry.DataHandler<D> dataHandler,
+                                              ExecutorService executorService){
+        assert chunkedJobDefinition !=null && getter!=null && dataHandler!=null;
+
+        executorService.execute(() -> {
+
+            Queue<Iterator<E>> chunksQueue = chunkedJobDefinition.getDataChunks();
+
+            if(!chunkedJobDefinition.tryLockProcessing()){
+                return;
+            }
+
+            try {
+                if(chunkedJobDefinition.getResult().isDone()){
+                    return;
+                }
+
+                Iterator<E> elements = chunksQueue.isEmpty() ? Collections.emptyIterator() : chunksQueue.remove();
+
+                EntryListComputationContext context = chunkedJobDefinition.getProcessedDataContext();
+
+                if(context == null){
+                    if(chunkedJobDefinition.isLastChunkRetrieved() && chunksQueue.isEmpty() && !elements.hasNext()){
+                        chunkedJobDefinition.complete(EMPTY_LIST);
+                        return;
+                    }
+                    context = generateComputationContext();
+                    chunkedJobDefinition.setProcessedDataContext(context);
+                }
+
+                do {
+                    if(elements.hasNext()){
+                        applyElementsComputation(elements, getter, dataHandler, context);
+                    }
+                    if(chunksQueue.isEmpty()){
+                        break;
+                    }
+                    elements = chunksQueue.remove();
+                } while (true);
+
+                if(chunkedJobDefinition.isLastChunkRetrieved() && chunksQueue.isEmpty()){
+                    chunkedJobDefinition.complete(context.metadataSchema == null ? EMPTY_LIST : convert(context));
+                }
+
+            } catch (Throwable throwable){
+                chunkedJobDefinition.getResult().completeExceptionally(throwable);
+                return;
+            } finally {
+                chunkedJobDefinition.unlockProcessing();
+            }
+
+            if(!chunksQueue.isEmpty() || chunkedJobDefinition.isLastChunkRetrieved() && !chunkedJobDefinition.getResult().isDone()){
```

Comment:
If I understand it correctly, `chunkedJobDefinition.isLastChunkRetrieved() && !chunkedJobDefinition.getResult().isDone()` AND the fact that `lastChunkRetrieved` is volatile, is to avoid an edge case where the last (few) chunk(s) never get processed.

This is very subtle and not straightforward (of course, multi-threading programming is almost never straightforward!) to understand. Would be great if you could add some comments.

Reply:
I believe I owe explanations about all the checks in this method because many of them are not obvious at the first look. Even so it's better to write self-explanatory code - it's complected to write self-explanatory code for multi-threading parts. Thus, I believe we should have more comments in this method, so that future developers don't get lost in it :sweat_smile: 
I tried to add detailed comments to each of the case to show the reasoning behind each of the case. Please, let me know what you think about it and it you have any concerns about any of those checks / comments. 
Commit with the comments: https://github.com/JanusGraph/janusgraph/pull/3760/commits/e5902ad541ba2d359a51e138ade24da717862bcc

Reply:
Perfect, it makes sense now.